### ISSUE TITLE
SIGILL: illegal instruction on old AMD Athlon(tm) II X2 245

### ISSUE THREAD

**Issue by iwanbk** [klauspost/reedsolomon#4]
First, thanks for this fast library.
It worked fine on my Intel CPU but crashed on my old AMD CPU.

I know it is old, just in case similar processor also have this issue.

This is error message from `go test`

```
$ go test ./...
SIGILL: illegal instruction
PC=0x453ac1

goroutine 26 [running]:
runtime: unknown argument frame size for github.com/klauspost/reedsolomon.galMulSSE3 called from 0x443127 [github.com/klauspost/reedsolomon.galMulSlice]
github.com/klauspost/reedsolomon.galMulSSE3()
	/home/ibk/pathgo/src/github.com/klauspost/reedsolomon/galois_amd64.s:57 +0x31 fp=0x7fdce8e87c50 sp=0x7fdce8e87c48
github.com/klauspost/reedsolomon.galMulSlice(0x4ff919, 0xc208026040, 0x12, 0x12, 0xc208026060, 0x12, 0x12)
	/home/ibk/pathgo/src/github.com/klauspost/reedsolomon/galois_amd64.go:37 +0x157 fp=0x7fdce8e87e48 sp=0x7fdce8e87c50
github.com/klauspost/reedsolomon.TestGalois(0xc20804e3f0)
	/home/ibk/pathgo/src/github.com/klauspost/reedsolomon/galois_test.go:134 +0x3d1 fp=0x7fdce8e87f68 sp=0x7fdce8e87e48
testing.tRunner(0xc20804e3f0, 0x648f70)
	/home/ibk/go/src/pkg/testing/testing.go:422 +0x8b fp=0x7fdce8e87f98 sp=0x7fdce8e87f68
runtime.goexit()
	/home/ibk/go/src/pkg/runtime/proc.c:1445 fp=0x7fdce8e87fa0 sp=0x7fdce8e87f98
created by testing.RunTests
	/home/ibk/go/src/pkg/testing/testing.go:504 +0x8db

goroutine 16 [chan receive]:
testing.RunTests(0x5a8500, 0x648ee0, 0x12, 0x12, 0x1)
	/home/ibk/go/src/pkg/testing/testing.go:505 +0x923
testing.Main(0x5a8500, 0x648ee0, 0x12, 0x12, 0x648a20, 0x10, 0x10, 0x647d80, 0x3, 0x3)
	/home/ibk/go/src/pkg/testing/testing.go:435 +0x84
main.main()
	github.com/klauspost/reedsolomon/_test/_testmain.go:119 +0x9c

goroutine 19 [finalizer wait]:
runtime.park(0x413290, 0x64d590, 0x64c0a9)
	/home/ibk/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x64d590, 0x64c0a9)
	/home/ibk/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
	/home/ibk/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
	/home/ibk/go/src/pkg/runtime/proc.c:1445

rax     0x10
rbx     0xf
rcx     0x8000001b00000005
rdx     0xc208026060
rdi     0x7fdce8e87e40
rsi     0xc208026040
rbp     0x10
rsp     0x7fdce8e87c48
r8      0x19
r9      0x12
r10     0x12
r11     0xc208026060
r12     0xc208026040
r13     0xc2080420f0
r14     0x2
r15     0xc2080004e0
rip     0x453ac1
rflags  0x10202
cs      0x33
fs      0x0
gs      0x0
FAIL	github.com/klauspost/reedsolomon	0.230s
```

And this is my cpuinfo
```
$ cat /proc/cpuinfo 
processor	: 0
vendor_id	: AuthenticAMD
cpu family	: 16
model		: 6
model name	: AMD Athlon(tm) II X2 245 Processor
stepping	: 2
microcode	: 0x10000b7
cpu MHz		: 800.000
cache size	: 1024 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 2
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 5
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm 3dnowext 3dnow constant_tsc rep_good nopl nonstop_tsc extd_apicid pni monitor cx16 popcnt lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt hw_pstate npt lbrv svm_lock nrip_save vmmcall
bogomips	: 5826.11
TLB size	: 1024 4K pages
clflush size	: 64
cache_alignment	: 64
address sizes	: 48 bits physical, 48 bits virtual
power management: ts ttp tm stc 100mhzsteps hwpstate

processor	: 1
vendor_id	: AuthenticAMD
cpu family	: 16
model		: 6
model name	: AMD Athlon(tm) II X2 245 Processor
stepping	: 2
microcode	: 0x10000b7
cpu MHz		: 1700.000
cache size	: 1024 KB
physical id	: 0
siblings	: 2
core id		: 1
cpu cores	: 2
apicid		: 1
initial apicid	: 1
fpu		: yes
fpu_exception	: yes
cpuid level	: 5
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm 3dnowext 3dnow constant_tsc rep_good nopl nonstop_tsc extd_apicid pni monitor cx16 popcnt lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt hw_pstate npt lbrv svm_lock nrip_save vmmcall
bogomips	: 5826.11
TLB size	: 1024 4K pages
clflush size	: 64
cache_alignment	: 64
address sizes	: 48 bits physical, 48 bits virtual
power management: ts ttp tm stc 100mhzsteps hwpstate
```

**Comment by klauspost** [comment]
Very strange. It obviously doesn't have SSE3, but it is detected as having it.

Could you go to `your-gopath/github.com/klauspost/cpuid` and capture the output of `go test -v`?

**Comment by iwanbk** [comment]
@klauspost 
here are the result
```
$ go test -v 
=== RUN TestCPUID
--- PASS: TestCPUID (0.00 seconds)
	cpuid_test.go:14: Max Function:0x5
	cpuid_test.go:16: Max Extended Function:0x8000001b
	cpuid_test.go:19: CPUID.(EAX=07H):EBX: 0x0
	cpuid_test.go:21: Name: AMD Athlon(tm) II X2 245 Processor
	cpuid_test.go:22: PhysicalCores: 2
	cpuid_test.go:23: ThreadsPerCore: 1
	cpuid_test.go:24: LogicalCores: 2
	cpuid_test.go:25: Family 16 Model: 6
	cpuid_test.go:26: Features: CMOV,NX,AMD3DNOW,AMD3DNOWEXT,MMX,MMXEXT,SSE,SSE2,SSE3,,LZCNT,POPCNT,RDTSCP,CX16
	cpuid_test.go:27: Cacheline bytes: 64
	cpuid_test.go:29: Yay - we have SSE 2
=== RUN TestCmov
--- PASS: TestCmov (0.00 seconds)
	cpuid_test.go:58: CMOV Support: true
=== RUN TestAmd3dnow
--- PASS: TestAmd3dnow (0.00 seconds)
	cpuid_test.go:68: AMD3DNOW Support: true
=== RUN TestAmd3dnowExt
--- PASS: TestAmd3dnowExt (0.00 seconds)
	cpuid_test.go:78: AMD3DNOWEXT Support: true
=== RUN TestMMX
--- PASS: TestMMX (0.00 seconds)
	cpuid_test.go:88: MMX Support: true
=== RUN TestMMXext
--- PASS: TestMMXext (0.00 seconds)
	cpuid_test.go:98: MMXEXT Support: true
=== RUN TestSSE
--- PASS: TestSSE (0.00 seconds)
	cpuid_test.go:108: SSE Support: true
=== RUN TestSSE2
--- PASS: TestSSE2 (0.00 seconds)
	cpuid_test.go:118: SSE2 Support: true
=== RUN TestSSE3
--- PASS: TestSSE3 (0.00 seconds)
	cpuid_test.go:128: SSE3 Support: true
=== RUN TestSSSE3
--- PASS: TestSSSE3 (0.00 seconds)
	cpuid_test.go:138: SSSE3 Support: false
=== RUN TestSSE4
--- PASS: TestSSE4 (0.00 seconds)
	cpuid_test.go:148: SSE4 Support: false
=== RUN TestSSE42
--- PASS: TestSSE42 (0.00 seconds)
	cpuid_test.go:158: SSE42 Support: false
=== RUN TestAVX
--- PASS: TestAVX (0.00 seconds)
	cpuid_test.go:168: AVX Support: false
=== RUN TestAVX2
--- PASS: TestAVX2 (0.00 seconds)
	cpuid_test.go:178: AVX2 Support: false
=== RUN TestFMA3
--- PASS: TestFMA3 (0.00 seconds)
	cpuid_test.go:188: FMA3 Support: false
=== RUN TestFMA4
--- PASS: TestFMA4 (0.00 seconds)
	cpuid_test.go:198: FMA4 Support: false
=== RUN TestXOP
--- PASS: TestXOP (0.00 seconds)
	cpuid_test.go:208: XOP Support: false
=== RUN TestF16C
--- PASS: TestF16C (0.00 seconds)
	cpuid_test.go:218: F16C Support: false
=== RUN TestCX16
--- PASS: TestCX16 (0.00 seconds)
	cpuid_test.go:228: CX16 Support: true
=== RUN TestBMI1
--- PASS: TestBMI1 (0.00 seconds)
	cpuid_test.go:238: BMI1 Support: false
=== RUN TestBMI2
--- PASS: TestBMI2 (0.00 seconds)
	cpuid_test.go:248: BMI2 Support: false
=== RUN TestTBM
--- PASS: TestTBM (0.00 seconds)
	cpuid_test.go:258: TBM Support: false
=== RUN TestLzcnt
--- PASS: TestLzcnt (0.00 seconds)
	cpuid_test.go:268: LZCNT Support: true
=== RUN TestPopcnt
--- PASS: TestPopcnt (0.00 seconds)
	cpuid_test.go:278: POPCNT Support: true
=== RUN TestAesNi
--- PASS: TestAesNi (0.00 seconds)
	cpuid_test.go:288: AESNI Support: false
=== RUN TestHTT
--- PASS: TestHTT (0.00 seconds)
	cpuid_test.go:298: HTT Support: false
=== RUN TestClmul
--- PASS: TestClmul (0.00 seconds)
	cpuid_test.go:308: CLMUL Support: false
=== RUN TestSSE2Slow
--- PASS: TestSSE2Slow (0.00 seconds)
	cpuid_test.go:318: SSE2SLOW Support: false
=== RUN TestSSE3Slow
--- PASS: TestSSE3Slow (0.00 seconds)
	cpuid_test.go:328: SSE3SLOW Support: false
=== RUN TestAtom
--- PASS: TestAtom (0.00 seconds)
	cpuid_test.go:338: ATOM Support: false
=== RUN TestNX
--- PASS: TestNX (0.00 seconds)
	cpuid_test.go:348: NX Support: true
=== RUN TestSSE4A
--- PASS: TestSSE4A (0.00 seconds)
	cpuid_test.go:358: SSE4A Support: true
=== RUN TestHLE
--- PASS: TestHLE (0.00 seconds)
	cpuid_test.go:368: HLE Support: false
=== RUN TestRTM
--- PASS: TestRTM (0.00 seconds)
	cpuid_test.go:378: RTM Support: false
=== RUN TestRdrand
--- PASS: TestRdrand (0.00 seconds)
	cpuid_test.go:388: Rdrand Support: false
=== RUN TestRdseed
--- PASS: TestRdseed (0.00 seconds)
	cpuid_test.go:398: Rdseed Support: false
=== RUN TestADX
--- PASS: TestADX (0.00 seconds)
	cpuid_test.go:408: ADX Support: false
=== RUN TestSHA
--- PASS: TestSHA (0.00 seconds)
	cpuid_test.go:418: SHA Support: false
=== RUN TestAVX512F
--- PASS: TestAVX512F (0.00 seconds)
	cpuid_test.go:428: AVX512F Support: false
=== RUN TestAVX512DQ
--- PASS: TestAVX512DQ (0.00 seconds)
	cpuid_test.go:438: AVX512DQ Support: false
=== RUN TestAVX512IFMA
--- PASS: TestAVX512IFMA (0.00 seconds)
	cpuid_test.go:448: AVX512IFMA Support: false
=== RUN TestAVX512PF
--- PASS: TestAVX512PF (0.00 seconds)
	cpuid_test.go:458: AVX512PF Support: false
=== RUN TestAVX512ER
--- PASS: TestAVX512ER (0.00 seconds)
	cpuid_test.go:468: AVX512ER Support: false
=== RUN TestAVX512CD
--- PASS: TestAVX512CD (0.00 seconds)
	cpuid_test.go:478: AVX512CD Support: false
=== RUN TestAVX512BW
--- PASS: TestAVX512BW (0.00 seconds)
	cpuid_test.go:488: AVX512BW Support: false
=== RUN TestAVX512VL
--- PASS: TestAVX512VL (0.00 seconds)
	cpuid_test.go:498: AVX512VL Support: false
=== RUN TestAVX512VBMI
--- PASS: TestAVX512VBMI (0.00 seconds)
	cpuid_test.go:508: AVX512VBMI Support: false
=== RUN TestMPX
--- PASS: TestMPX (0.00 seconds)
	cpuid_test.go:518: MPX Support: false
=== RUN TestERMS
--- PASS: TestERMS (0.00 seconds)
	cpuid_test.go:528: ERMS Support: false
=== RUN TestVendor
--- PASS: TestVendor (0.00 seconds)
	cpuid_test.go:533: Vendor ID: 2
=== RUN TestIntel
--- PASS: TestIntel (0.00 seconds)
	cpuid_test.go:543: TestIntel: false
=== RUN TestAMD
--- PASS: TestAMD (0.00 seconds)
	cpuid_test.go:553: TestAMD: true
=== RUN TestTransmeta
--- PASS: TestTransmeta (0.00 seconds)
	cpuid_test.go:563: TestTransmeta: false
=== RUN TestNSC
--- PASS: TestNSC (0.00 seconds)
	cpuid_test.go:573: TestNSC: false
=== RUN TestVIA
--- PASS: TestVIA (0.00 seconds)
	cpuid_test.go:583: TestVIA: false
=== RUN TestVM
--- PASS: TestVM (0.00 seconds)
	cpuid_test.go:588: Vendor ID: false
=== RUN TestRtCounter
--- PASS: TestRtCounter (0.00 seconds)
	cpuid_test.go:595: CPU Counter: 6712892505249 6712892505423 174
=== RUN TestIa32TscAux
--- PASS: TestIa32TscAux (0.00 seconds)
	cpuid_test.go:601: Ia32TscAux:0x1
	cpuid_test.go:605: Likely chip, core: 0 1
=== RUN TestLogicalCPU
--- PASS: TestLogicalCPU (0.00 seconds)
	cpuid_test.go:611: Currently executing on cpu: 1
=== RUN TestMaxFunction
--- PASS: TestMaxFunction (0.00 seconds)
PASS
ok  	github.com/klauspost/cpuid	0.004s
```

**Comment by klauspost** [comment]
Very weird. It *does* actually SSE3 ([confirmed here](https://en.wikipedia.org/wiki/List_of_AMD_Athlon_X2_microprocessors)). I will check if I have accidentally used an instruction from SSE4 or something else.

**Comment by klauspost** [comment]
Ah- I'm stupid. All this time i assumed it was SSE3, but PSHUFB is S(upplemental)SSE3 :/

Fixed in commit cfcc65d3875962847885b80a37de0864890a268a

**Comment by klauspost** [comment]
Closing. Hope it helps! Thanks again for reporting!
